2.4/root/usr/share/container-scripts/mongodb/common.sh

```diff
@@ -117,23 +117,60 @@ function mongo_initiate() {
 
 # get the address of the current primary member
 function mongo_primary_member_addr() {
-  local current_endpoints=$(endpoints)
-  local mongo_node="$(echo "${current_endpoints}" | grep -v "$(container_addr)" | head -1):${CONTAINER_PORT}"
-  echo -n $(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host ${mongo_node} --quiet --eval "print(rs.isMaster().primary);")
+  endpoints | grep -v "$(container_addr)" | while read mongo_node; do
+    cmd_output="$(mongo admin -u admin -p "$MONGODB_ADMIN_PASSWORD" --host "$mongo_node:$CONTAINER_PORT" --eval 'print(rs.isMaster().primary)' --quiet || true)"
+
+    # Trying to find IP:PORT in output and filter out error message because mongo prints it to stdout
+    ip_and_port_regexp='[0-9]\+\.[0-9]\+\.[0-9]\+\.[0-9]\+:[0-9]\+'
+    if addr="$(echo "$cmd_output" | grep -x "$ip_and_port_regexp")"; then
+      echo "$addr"
+      exit 0
+    fi
+
+    echo >&2 "Cannot get address of primary from $mongo_node node: $cmd_output"
+    exit 1
+  done && return
+
+  echo >&2 "Cannot get address of primary node: after checking all nodes we don't have the address"
+
+  # return empty string
+  echo
 }
 
 # mongo_remove removes the current MongoDB from the cluster
 function mongo_remove() {
-  echo "=> Removing $(mongo_addr) on $(mongo_primary_member_addr) ..."
+  local primary_addr
+  primary_addr="$(mongo_primary_member_addr)"
+
+  if [ -z "$primary_addr" ]; then
+    echo >&2 "Cannot remove node from the replica set: primary node was not found"
```

Comment:
I agree with you... maybe there's a way we can get best of both worlds?

Wrapping every error with a message is probably over kill, and not having `set -e` and ignoring errors silently is not a great idea.

In this particular case here, can't we simply make `mongo_primary_member_addr` print an error message (it already does)? And then it would print an error message and exit in case of errors, not requiring callers to handle empty strings.